File: javascript/cote/LTCD/조합의_합.mjs

const TEST_CASES = 10
const MAX_VALUE = 100
const N = 15

function solve(candidates, target) {
	const ret = []

	function dfs(cur, path, acc) {

		if (candidates.length <= cur) {
			return
		}

		if (acc > target) {
			return
		}

		if (acc == target) {
			ret.push([...path])
			return
		}

		for (let i = cur; i < candidates.length; i++) {
			path.push(candidates[i])
			dfs(i + 1, path, acc + candidates[i])
			path.pop()
		}

		return
	}

	dfs(0, [], 0)

	return ret
}

for (let it = 0; it < TEST_CASES; it++) {
	
	const candidates = []

	for (let i = 0; i < N; i++) {
		candidates.push(Math.floor(Math.random() * MAX_VALUE))
	}

	const target = Math.floor(Math.random() * MAX_VALUE * 1.5)

	console.log(solve(candidates, target))
} 

File: python/cote/LTCD/조합의_합.py
from typing import List
import random

TEST_CASES = 10
N = 15
MAX_VALUE = 100


def solve(candidates: List[int], target: int) -> List[List[int]]:
    ret = []

    def dfs(path: List[int], start: int, acc: int):

        if acc > target:
            return

        if acc == target:
            ret.append(path)
            return

        for i in range(start, len(candidates)):
            dfs([*path, candidates[i]], i + 1, acc + candidates[i])

    dfs([], 0, 0)

    return ret


if __name__ == "__main__":
    random.seed(43)

    for _ in range(TEST_CASES):
        candidates = [random.randrange(MAX_VALUE) for _ in range(N)]
        target = random.randrange(MAX_VALUE * 1.5)
        print(solve(candidates, target))
